syntax = "proto3";

// import "google/protobuf/empty.proto";

package simulator_api;

service Simulator {
  rpc DoResourceAction(ResourceActionMessage) returns (CommandStatusMessage) {}
  rpc CreateVM(CreateVMMessage) returns (CommandStatusMessage) {}
  rpc DoVMAction(VMActionMessage) returns (CommandStatusMessage) {}

  // event-loop commands
  rpc SimulateAll(EmptyMessage/*google.protobuf.Empty*/) returns (stream LogMessage) {}
}

message EmptyMessage {

}

enum ResourceActionType {
  UNSPECIFIED_RESOURCE_ACTION = 0;
  BOOT_RESOURCE_ACTION = 1;
  REBOOT_RESOURCE_ACTION = 2;
  SHUTDOWN_RESOURCE_ACTION = 3;
}

enum VMActionType {
  UNSPECIFIED_VM_ACTION = 0;
  PROVISION_VM_ACTION = 1;
  REBOOT_VM_ACTION = 2;
  STOP_VM_ACTION = 3;
  DELETE_VM_ACTION = 4;
}

message CommandStatusMessage {
  string error_text = 3;
}

message ResourceActionMessage {
  string resource_name = 1;
  ResourceActionType resource_action_type = 4;
}

message VMActionMessage {
  string vm_name = 1;
  VMActionType vm_action_type = 4;
}

// temporal! a researcher should provide his custom workload spec manually for now
message CreateVMMessage {
  string vm_name = 1;
  uint32 required_ram = 2;
}

enum LogSeverity {
  UNSPECIFIED_LOG_SEVERITY = 0;
  ERROR_LOG_SEVERITY = 1;
  INFO_LOG_SEVERITY = 2;
  DEBUG_LOG_SEVERITY = 3;
}

message LogMessage {
  bool is_last = 10;
  uint64 time = 9;
  string caller_type = 5;
  string caller_name = 6;
  LogSeverity severity = 7;
  string text = 8;
}
